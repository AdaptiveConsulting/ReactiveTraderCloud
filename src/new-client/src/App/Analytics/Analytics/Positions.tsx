import React from "react"
import { BubbleChart, Title } from "./styled"
// import PositionsBubbleChart from './positions-chart/PositionsBubbleChart'

export const Positions: React.FC = () => {
  return (
    <div>
      <Title>Positions</Title>
      <BubbleChart>{/*<PositionsBubbleChart />*/}</BubbleChart>
    </div>
  )
}
