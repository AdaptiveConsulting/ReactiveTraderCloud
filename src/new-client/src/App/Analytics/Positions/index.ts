export { Positions, positions$ } from "./Positions"
