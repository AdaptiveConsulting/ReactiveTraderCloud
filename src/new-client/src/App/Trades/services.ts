import { bind } from "@react-rxjs/core"
import { combineLatest, Subject } from "rxjs"
import { map, scan, startWith } from "rxjs/operators"
import { trades$ } from "services/trades"
import { ColField } from "./TradesGrid"

type SortDirection = "ASC" | "DESC"

class TableSort {
  constructor(public direction?: SortDirection, public field?: ColField) {}
}

export const sortFieldSelections$ = new Subject<ColField>()

const descDefaultFields = new Set<ColField>([
  "tradeDate",
  "valueDate",
  "tradeId",
])

export const [useTableSort, tableSort$] = bind(
  sortFieldSelections$.pipe(
    scan((tableSort, sortFieldSelection) => {
      if (tableSort.field === sortFieldSelection) {
        if (
          descDefaultFields.has(sortFieldSelection) &&
          tableSort.direction === "DESC"
        ) {
          return new TableSort("ASC", sortFieldSelection)
        } else if (
          !descDefaultFields.has(sortFieldSelection) &&
          tableSort.direction === "ASC"
        ) {
          return new TableSort("DESC", sortFieldSelection)
        } else {
          return new TableSort()
        }
      }

      return new TableSort(
        descDefaultFields.has(sortFieldSelection) ? "DESC" : "ASC",
        sortFieldSelection,
      )
    }, new TableSort()),
    startWith(new TableSort()),
  ),
)

const numericComparator = (direction: SortDirection, a: number, b: number) => {
  return direction === "ASC" ? a - b : b - a
}

const stringComparator = (direction: SortDirection, a: string, b: string) => {
  if (a > b) {
    return direction === "ASC" ? 1 : -1
  } else if (a < b) {
    return direction === "ASC" ? -1 : 1
  } else {
    return 0
  }
}

export const [useTableTrades, tableTrades$] = bind(
  combineLatest([trades$, tableSort$]).pipe(
    map(([trades, { field, direction }]) => {
      const sortedTrades = [...trades]
      if (field && direction) {
        sortedTrades.sort((tradeA, tradeB) => {
          const aSortFieldValue = tradeA[field]
          const bSortFieldValue = tradeB[field]

          if (
            aSortFieldValue instanceof Date &&
            bSortFieldValue instanceof Date
          ) {
            return numericComparator(
              direction,
              aSortFieldValue.valueOf(),
              bSortFieldValue.valueOf(),
            )
          }

          if (
            typeof aSortFieldValue === "number" &&
            typeof bSortFieldValue === "number"
          ) {
            return numericComparator(
              direction,
              aSortFieldValue,
              bSortFieldValue,
            )
          }

          if (
            typeof aSortFieldValue === "string" &&
            typeof bSortFieldValue === "string"
          ) {
            return stringComparator(
              direction,
              aSortFieldValue.toLowerCase(),
              bSortFieldValue.toLowerCase(),
            )
          }

          throw new Error(
            "Trade sort for table received value of unexpected type",
          )
        })
      }
      return sortedTrades
    }),
  ),
)
