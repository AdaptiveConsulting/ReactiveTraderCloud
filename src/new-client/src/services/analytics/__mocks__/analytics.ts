import { bind } from "@react-rxjs/core"
import { defer, of, Observable } from "rxjs"
import { HistoryEntry, CurrencyPairPosition } from "../types"

let historyMocks$: Observable<HistoryEntry[]> = of([])
let positionMocks$: Observable<Record<string, CurrencyPairPosition>> = of({})

export const __setHistoryMock = (value: Observable<HistoryEntry[]>) => {
  historyMocks$ = value
}

export const __resetHistoryMocks = () => {
  historyMocks$ = of([])
}

export const history$ = defer(() => historyMocks$)

export const __setPositionMock = (
  value: Observable<Record<string, CurrencyPairPosition>>,
) => {
  positionMocks$ = value
}

export const __resetPositionMocks = () => {
  positionMocks$ = of({})
}

export const [useCurrentPositions, currentPositions$] = bind(
  defer(() => {
    return positionMocks$
  }),
)

export const __resetMocks = () => {
  __resetHistoryMocks()
  __resetPositionMocks()
}
