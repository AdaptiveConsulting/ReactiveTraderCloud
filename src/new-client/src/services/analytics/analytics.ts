import { bind, shareLatest } from "@react-rxjs/core"
import { Observable } from "rxjs"
import { map } from "rxjs/operators"
import { getStream$ } from "../client"
import { CurrencyPairPosition, HistoryEntry, PositionsRaw } from "./types"

export const analytics$ = getStream$<PositionsRaw, string>(
  "analytics",
  "getAnalytics",
  "USD",
).pipe(shareLatest())

export const history$: Observable<HistoryEntry[]> = analytics$.pipe(
  map((analitics) =>
    analitics.History.filter(Boolean).map(({ Timestamp, UsdPnl }) => ({
      usPnl: UsdPnl,
      timestamp: new Date(Timestamp).getTime(),
    })),
  ),
  shareLatest(),
)

export const [useCurrentPositions, currentPositions$] = bind<
  Record<string, CurrencyPairPosition>
>(
  analytics$.pipe(
    map((analitics) =>
      Object.fromEntries(
        analitics.CurrentPositions.map(
          ({ Symbol, BasePnl, BaseTradedAmount, CounterTradedAmount }) => [
            Symbol,
            {
              symbol: Symbol,
              basePnl: BasePnl,
              baseTradedAmount: BaseTradedAmount,
              counterTradedAmount: CounterTradedAmount,
              basePnlName: "basePnl",
              baseTradedAmountName: "baseTradedAmount",
            },
          ],
        ),
      ),
    ),
  ),
)
