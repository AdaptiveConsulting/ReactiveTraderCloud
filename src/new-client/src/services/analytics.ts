import { bind, shareLatest } from "@react-rxjs/core"
import { map } from "rxjs/operators"
import { CamelCase } from "services/utils"
import { getStream$ } from "./client"

interface CurrencyPairPositionRaw {
  Symbol: string
  BasePnl: number
  BaseTradedAmount: number
  CounterTradedAmount: number
}

export interface CurrencyPairPosition
  extends CamelCase<CurrencyPairPositionRaw> {
  basePnlName: "basePnl"
  baseTradedAmountName: "baseTradedAmount"
}

interface HistoryRaw {
  Timestamp: string
  UsdPnl: number
}
export interface HistoryEntry {
  timestamp: number
  usPnl: number
}

interface PositionsRaw {
  CurrentPositions: CurrencyPairPositionRaw[]
  History: HistoryRaw[]
}

const analytics$ = getStream$<PositionsRaw, string>(
  "analytics",
  "getAnalytics",
  "USD",
).pipe(shareLatest())

export const [useHistory, history$] = bind<HistoryEntry[]>(
  analytics$.pipe(
    map((analitics) =>
      analitics.History.filter(Boolean).map(({ Timestamp, UsdPnl }) => ({
        usPnl: UsdPnl,
        timestamp: new Date(Timestamp).getTime(),
      })),
    ),
  ),
)

export const [useCurrentPositions, currentPositions$] = bind<
  Record<string, CurrencyPairPosition>
>(
  analytics$.pipe(
    map((analitics) =>
      Object.fromEntries(
        analitics.CurrentPositions.map(
          ({ Symbol, BasePnl, BaseTradedAmount, CounterTradedAmount }) => [
            Symbol,
            {
              symbol: Symbol,
              basePnl: BasePnl,
              baseTradedAmount: BaseTradedAmount,
              counterTradedAmount: CounterTradedAmount,
              basePnlName: "basePnl",
              baseTradedAmountName: "baseTradedAmount",
            },
          ],
        ),
      ),
    ),
  ),
)
