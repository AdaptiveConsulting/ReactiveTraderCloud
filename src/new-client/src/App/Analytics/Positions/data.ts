import { scaleLinear, SimulationNodeDatum } from "d3"
import { combineLatest, Observable } from "rxjs"
import {
  distinctUntilChanged,
  filter,
  map,
  scan,
  withLatestFrom,
} from "rxjs/operators"
import { colors } from "@/theme"
import { currentPositions$ } from "@/services/analytics"
import { currencyPairs$ } from "@/services/currencyPairs"
import { formatAsWholeNumber } from "@/utils/formatNumber"
import { bind, shareLatest } from "@react-rxjs/core"
import { mapObject } from "@/utils/mapObject"
import { equals } from "@/utils/equals"

interface CCYPosition {
  symbol: string
  baseTradedAmount: number
}

export interface BubbleChartNode extends SimulationNodeDatum {
  id: string
  r: number
  color: string
  text: string
}

export const [useData, data$] = bind(
  combineLatest([currencyPairs$, currentPositions$]).pipe(
    filter((_, idx) => idx === 0),
  ),
)

export const nodes$: Observable<BubbleChartNode[]> = currentPositions$.pipe(
  map((positions) =>
    mapObject(positions, ({ baseTradedAmount, counterTradedAmount }) => ({
      baseTradedAmount,
      counterTradedAmount,
    })),
  ),
  distinctUntilChanged(equals),
  withLatestFrom(currencyPairs$),
  map(([positions, currencyPairs]) => {
    const aggregatedPositions = Object.entries(positions).reduce(
      (acc, [symbol, { baseTradedAmount, counterTradedAmount }]) => {
        const ccyPair = currencyPairs[symbol]
        const baseCurrency = ccyPair ? ccyPair.base : ""
        const counterCurrency = ccyPair ? ccyPair.terms : ""
        acc[baseCurrency] = (acc[baseCurrency] || 0) + baseTradedAmount
        acc[counterCurrency] = (acc[counterCurrency] || 0) + counterTradedAmount
        return acc
      },
      {} as Record<string, number>,
    )
    return Object.entries(aggregatedPositions)
      .map(([key, val]) => ({
        symbol: key,
        baseTradedAmount: val,
      }))
      .filter((posPerCCY: CCYPosition) => posPerCCY.baseTradedAmount !== 0)
  }),
  map((positionData) => {
    const minR: number = 15
    const maxR: number = 60
    const baseValues: number[] = positionData.map((val) =>
      Math.abs(val.baseTradedAmount),
    )
    const maxValue: number = Math.max(...baseValues) || 0
    const minValue: number =
      Math.min(...baseValues) !== maxValue ? Math.min(...baseValues) : 0
    return [
      positionData,
      {
        r: scaleLinear().domain([minValue, maxValue]).range([minR, maxR]),
      },
    ] as const
  }),
  scan((acc, [positionsData, scales]) => {
    // start from scratch with a new object so deleted nodes are not silently included
    const newAcc: Record<string, BubbleChartNode> = {}
    positionsData.forEach((dataObj: CCYPosition) => {
      const color =
        dataObj.baseTradedAmount > 0
          ? colors.accents.positive.base
          : colors.accents.negative.base

      newAcc[dataObj.symbol] = Object.assign(
        acc[dataObj.symbol] || { id: dataObj.symbol },
        {
          color,
          r: scales.r(Math.abs(dataObj.baseTradedAmount)),
          text: formatAsWholeNumber(dataObj.baseTradedAmount),
        },
      )
    })
    return newAcc
  }, {} as Record<string, BubbleChartNode>),
  map(Object.values),
  shareLatest(),
)
