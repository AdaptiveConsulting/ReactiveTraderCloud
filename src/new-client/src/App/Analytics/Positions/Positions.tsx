import { useLayoutEffect, useRef } from "react"
import {
  D3DragEvent,
  drag,
  forceCollide,
  forceSimulation,
  forceX,
  forceY,
  select,
} from "d3"
import { BubbleChartNode, nodes$, useData, data$ } from "./data"
import { BubbleChart, Title } from "../styled"

// extra pixel amount that nodes in the chart repel each other within
// for collision detection purposes, a nodes radius is r + COLLIDE_BORDER_WIDTH pixels
const COLLIDE_BORDER_WIDTH = 2
const getId = (x: any) => x.id

const d3Effect = (chartDiv: HTMLDivElement) => {
  const tooltip = select(chartDiv)
    .append("div")
    .style("visibility", "hidden")
    .attr("class", "analytics__positions-tooltip")
    .attr("data-testid", "tooltip")

  // const svg: Selection<SVGElement> = select(chartDiv).select("svg")
  const { width, height } = chartDiv
    ? chartDiv.getBoundingClientRect()
    : { width: 0, height: 0 }
  const svg = select(chartDiv)
    .append("svg")
    .attr("width", width)
    .attr("height", height)

  // Add shadow
  const definitions = svg.append("defs")
  const filter = definitions
    .append("filter")
    .attr("id", "drop-shadow")
    .attr("height", "130%")

  filter
    .append("feGaussianBlur")
    .attr("in", "SourceAlpha")
    .attr("stdDeviation", 1.5)
    .attr("result", "blur")

  filter
    .append("feOffset")
    .attr("in", "blur")
    .attr("dx", 1)
    .attr("dy", 1)
    .attr("result", "offsetBlur")

  const feMerge = filter.append("feMerge")
  feMerge.append("feMergeNode").attr("in", "offsetBlur")
  feMerge.append("feMergeNode").attr("in", "SourceGraphic")

  const positionTooltip = (event: MouseEvent, node: BubbleChartNode): void => {
    if (typeof node.x === "undefined" || typeof node.y === "undefined") {
      return
    }
    const posX: number =
      (event ? event.offsetX : node.x) - tooltip.nodes()[0].clientWidth / 2
    const posY: number = event ? event.offsetY : node.y
    const id: string = node.id
    tooltip.style("top", posY + 15 + "px").style("left", posX + "px")
    tooltip.text(`${id} ${node.text}`)
  }

  const onMove = (
    event: D3DragEvent<any, BubbleChartNode, any>,
    d: BubbleChartNode,
  ) => {
    force.alpha(0.5).restart()
    positionTooltip(event.sourceEvent, d)
    d.fx = event.x
    d.fy = event.y
  }

  const force = forceSimulation<BubbleChartNode>()
    // forces that pull all nodes toward the center of the simulation
    .force(
      "forceX",
      forceX<BubbleChartNode>()
        .strength(0.1)
        .x(width * 0.5),
    )
    .force(
      "forceY",
      forceY<BubbleChartNode>()
        .strength(0.1)
        .y(height * 0.5),
    )
    // force that pushes nodes away from each other when they are
    // within each other's collision radius
    .force(
      "collide",
      forceCollide<BubbleChartNode>()
        .strength(0.5)
        .radius(function (d) {
          return d.r + COLLIDE_BORDER_WIDTH
        }),
    )
    .alphaMin(0.05)

  force.on("tick", () => {
    svg
      .selectAll("g:not(.exit)")
      .data(force.nodes(), getId)
      .join(
        (enter) =>
          enter
            .append("g")
            .style("visibility", "hidden")
            .call(
              drag<any, BubbleChartNode>()
                .on("start", onMove)
                .on("drag", onMove)
                .on("end", (event, d) => {
                  force.alphaTarget(0)
                  positionTooltip(event.sourceEvent, d)
                  d.fx = null
                  d.fy = null
                }),
            )
            .on("mouseover", (e, d) => {
              tooltip.style("visibility", "visible")
              positionTooltip(e, d)
            })
            .on("mousemove", positionTooltip)
            .on("mouseout", () => {
              tooltip.style("visibility", "hidden")
            })
            .attr("class", "node")
            .append("circle")
            .attr("r", (d) => d.r)
            .attr("cx", width / 2)
            .attr("cy", height / 2)
            .style("fill", (d: BubbleChartNode) => d.color)
            .style("filter", "url(#drop-shadow)")
            .select(function () {
              return this.parentNode as any
            })
            .append("text")
            .attr("text-anchor", "middle")
            .attr("class", "analytics__positions-label")
            .attr(
              "data-testid",
              (d: BubbleChartNode) => `positions-label-${d.id}`,
            )
            .text(getId),
        (update) =>
          update
            .style("visibility", "visible")
            .select("circle")
            .attr("cx", (d: any) => d.x)
            .attr("cy", (d: any) => d.y)
            .style("fill", (d: BubbleChartNode) => d.color)
            .attr("r", (d: BubbleChartNode) => d.r)
            .select(function () {
              return (this as any).parentNode as any
            })
            .select("text")
            .attr("x", (d: any) => d.x)
            .attr("y", (d: any) => d.y + 4),
        (exit) => exit.classed("exit", true).remove(),
      )
  })

  const subscription = nodes$.subscribe((nodes) => {
    force.restart().alpha(0.5).nodes(nodes)
  })

  return () => subscription.unsubscribe()
}

export const Positions: React.FC = () => {
  useData()
  const wrapperRef = useRef<HTMLDivElement>(null)
  useLayoutEffect(() => d3Effect(wrapperRef.current!), [])

  return (
    <div>
      <Title>Positions</Title>
      <BubbleChart ref={wrapperRef} />
    </div>
  )
}

export const positions$ = data$
