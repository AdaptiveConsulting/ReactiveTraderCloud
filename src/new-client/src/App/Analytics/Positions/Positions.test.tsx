import { Subscribe } from "@react-rxjs/core"
import { render, screen, act, fireEvent } from "@testing-library/react"
import { BehaviorSubject } from "rxjs"
import { TestThemeProvider } from "utils/testUtils"
import { Positions, positions$ } from "./Positions"
import { CurrencyPairPosition } from "services/analytics"
import { CurrencyPair } from "services/currencyPairs"
// import renderer from "react-test-renderer"

jest.mock("services/analytics/analytics")
jest.mock("services/currencyPairs/currencyPairs")

const currencyPairMock1: CurrencyPair = {
  symbol: "EURAUD",
  base: "EUR",
  terms: "AUD",
  ratePrecision: 5,
  pipsPosition: 4,
}

const currencyPairMock2: CurrencyPair = {
  symbol: "EURJPY",
  base: "EUR",
  terms: "JPY",
  ratePrecision: 5,
  pipsPosition: 2,
}

const positionMock: Record<string, CurrencyPairPosition> = {
  EURAUD: {
    symbol: "EURAUD",
    basePnl: 5505.700507776323,
    baseTradedAmount: 1000000,
    counterTradedAmount: -1557030,
    basePnlName: "basePnl",
    baseTradedAmountName: "baseTradedAmount",
  },
  EURJPY: {
    symbol: "EURJPY",
    basePnl: -28767.690451584618,
    baseTradedAmount: -7975223,
    counterTradedAmount: 1014984843.778,
    basePnlName: "basePnl",
    baseTradedAmountName: "baseTradedAmount",
  },
}

const positionMock2: Record<string, CurrencyPairPosition> = {
  EURAUD: {
    symbol: "EURAUD",
    basePnl: 5505.700507776323,
    baseTradedAmount: 1000000,
    counterTradedAmount: -1557031,
    basePnlName: "basePnl",
    baseTradedAmountName: "baseTradedAmount",
  },
  EURJPY: {
    symbol: "EURJPY",
    basePnl: -28767.690451584618,
    baseTradedAmount: -7975223,
    counterTradedAmount: 1014984843.778,
    basePnlName: "basePnl",
    baseTradedAmountName: "baseTradedAmount",
  },
}

const renderComponent = () =>
  render(
    <TestThemeProvider>
      <Subscribe source$={positions$} fallback="No data">
        <Positions />
      </Subscribe>
    </TestThemeProvider>,
  )

const _analytics = require("services/analytics/analytics")
const _ccpp = require("services/currencyPairs/currencyPairs")

describe("Positions", () => {
  beforeEach(() => {
    _analytics.__resetMocks()
    _ccpp.__resetMocks()
  })

  it("should render the correct price tag", () => {
    _ccpp.__setCurrencyPairsMock(currencyPairMock1.symbol, currencyPairMock1)
    _ccpp.__setCurrencyPairsMock(currencyPairMock2.symbol, currencyPairMock2)

    const positionMock$ = new BehaviorSubject<
      Record<string, CurrencyPairPosition>
    >(positionMock)
    _analytics.__setPositionMock(positionMock$)

    renderComponent()

    act(() => {
      fireEvent.mouseOver(screen.getByTestId("positions-label-JPY"))
    })

    expect(screen.getByTestId("tooltip").textContent).toBe("JPY 1,014,984,844")

    act(() => {
      fireEvent.mouseOver(screen.getByTestId("positions-label-AUD"))
    })

    expect(screen.getByTestId("tooltip").textContent).toBe("AUD -1,557,030")

    act(() => {
      positionMock$.next(positionMock2)
      fireEvent.mouseOver(screen.getByTestId("positions-label-AUD"))
    })

    expect(screen.getByTestId("tooltip").textContent).toBe("AUD -1,557,031")
  })

  it("should display the correct bubble chart", () => {
    _ccpp.__setCurrencyPairsMock(currencyPairMock1.symbol, currencyPairMock1)
    _ccpp.__setCurrencyPairsMock(currencyPairMock2.symbol, currencyPairMock2)

    const positionMock$ = new BehaviorSubject<
      Record<string, CurrencyPairPosition>
    >(positionMock)
    _analytics.__setPositionMock(positionMock$)

    const subscription = positions$.subscribe()
    const { container } = render(
      <TestThemeProvider>
        <Subscribe source$={positions$} fallback="No data">
          <Positions />
        </Subscribe>
      </TestThemeProvider>,
    )

    expect(container.firstChild).toMatchSnapshot()

    act(() => {
      positionMock$.next(positionMock2)
    })

    expect(container.firstChild).toMatchSnapshot()

    subscription.unsubscribe()
  })
})
