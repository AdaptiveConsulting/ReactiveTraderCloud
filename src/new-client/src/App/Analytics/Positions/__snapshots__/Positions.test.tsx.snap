// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Positions should display the correct bubble chart 1`] = `
.c0 {
  margin-bottom: 1rem;
  font-size: 15px;
  font-weight: normal;
}

.c1 {
  text-anchor: middle;
  height: 18rem;
  overflow: hidden;
  position: relative;
}

<div>
  <div
    class="c0"
  >
    Positions
  </div>
  <div
    class="c1"
  >
    <div
      class="analytics__positions-tooltip"
      data-testid="tooltip"
      style="visibility: hidden;"
    />
    <svg
      height="0"
      width="0"
    >
      <defs>
        <filter
          height="130%"
          id="drop-shadow"
        >
          <fegaussianblur
            in="SourceAlpha"
            result="blur"
            stdDeviation="1.5"
          />
          <feoffset
            dx="1"
            dy="1"
            in="blur"
            result="offsetBlur"
          />
          <femerge>
            <femergenode
              in="offsetBlur"
            />
            <femergenode
              in="SourceGraphic"
            />
          </femerge>
        </filter>
      </defs>
    </svg>
  </div>
</div>
`;

exports[`Positions should display the correct bubble chart 2`] = `
<div>
  <div
    class="sc-jSgupP dyNtTL"
  >
    Positions
  </div>
  <div
    class="sc-iBPRYJ ewjiun"
  >
    <div
      class="analytics__positions-tooltip"
      data-testid="tooltip"
      style="visibility: hidden;"
    />
    <svg
      height="0"
      width="0"
    >
      <defs>
        <filter
          height="130%"
          id="drop-shadow"
        >
          <fegaussianblur
            in="SourceAlpha"
            result="blur"
            stdDeviation="1.5"
          />
          <feoffset
            dx="1"
            dy="1"
            in="blur"
            result="offsetBlur"
          />
          <femerge>
            <femergenode
              in="offsetBlur"
            />
            <femergenode
              in="SourceGraphic"
            />
          </femerge>
        </filter>
      </defs>
      <g
        class="node"
        style="visibility: visible;"
      >
        <circle
          cx="22.366362534396142"
          cy="65.17120150536171"
          r="15.240588068460097"
          style="fill: #ff274b; filter: url(#drop-shadow);"
        />
        <text
          class="analytics__positions-label"
          data-testid="positions-label-EUR"
          text-anchor="middle"
          x="22.366362534396142"
          y="69.17120150536171"
        >
          EUR
        </text>
      </g>
      <g
        class="node"
        style="visibility: visible;"
      >
        <circle
          cx="-24.09298211184919"
          cy="64.28187563887715"
          r="15"
          style="fill: #ff274b; filter: url(#drop-shadow);"
        />
        <text
          class="analytics__positions-label"
          data-testid="positions-label-AUD"
          text-anchor="middle"
          x="-24.09298211184919"
          y="68.28187563887715"
        >
          AUD
        </text>
      </g>
      <g
        class="node"
        style="visibility: visible;"
      >
        <circle
          cx="0.13455356501548343"
          cy="-10.50968065293873"
          r="60"
          style="fill: #01c38d; filter: url(#drop-shadow);"
        />
        <text
          class="analytics__positions-label"
          data-testid="positions-label-JPY"
          text-anchor="middle"
          x="0.13455356501548343"
          y="-6.50968065293873"
        >
          JPY
        </text>
      </g>
    </svg>
  </div>
</div>
`;
